Add Contextualized Markov Graphs

cnellington opened the pull request:
Added contextualized markov graphs using neighborhood edge regression.
Added associated tests.

blengerich commented on contextualized/regression/trainers.py:
```diff
@@ -35,4 +35,18 @@ def predict_correlation(self, model, dataloader):
+        """
+        Returns context-specific precision matrix under a Gaussian graphical model
+        Assumes all diagonal precisions are equal and constant over context.
+        - omegas (numpy.ndarray): (n, x_dim, x_dim)
```
What is this? A return value?

blengerich commented on contextualized/regression/trainers.py:
```diff
@@ -35,4 +35,18 @@ def predict_correlation(self, model, dataloader):
+        - omegas (numpy.ndarray): (n, x_dim, x_dim)
+        """
+        betas, _ = super().predict_params(model, dataloader)
+        return -betas
```
Please document the equivalence between -beta and omega.

blengerich commented on contextualized/regression/tests.py:
```diff
@@ -56,29 +56,33 @@ def setUp(self):
         if correlation:
             rhos = trainer.predict_correlation(model, dataloader)
+        if markov:
+            omegas = trainer.predict_precision(model, dataloader)
```
Are the values returned here tested? Or is just testing that the `predict_` functions return without error?

cnellington replied:
Just to make sure there are no errors. It would be nice to have some test cases that check if we recover true models, but currently tests just check convergence in MSE and if any components throw errors.

cnellington commented:
Minor updates to NOTMAD to get better integration with existing tools and make it device agnostic, but another issue will tackle the NOTMAD integration. Updated Markov networks, added documentation, and fixed some cyclic import issues in regression.

cnellington commented:
Once this is merged we'll do the black/pylint style overhaul.

blengerich reviewed (approved):
LGTM. Great work, excited to use these Markov graphs.
